### DISCUSSION TITLE
How to add asynchronous commands to an extension?

### DISCUSSION THREAD

**Discussion by hans-pragt** [ueberdosis/tiptap#1630] [General Questions]
Hi! I am upgrading an extension from tiptap v1 to v2. I am running into a problem with one of its commands that looks like this:

```
addCommands() {
    return {
      asyncCommand: () => ({ state, dispatch }) => {
        if (!dispatch) {
          return false;
        }

        getAsyncData(state.doc)
        .then(data => {
          this.options.data = data;
          dispatch(state.tr);
        });

        return true;
      }
}
```

The data is then used by a custom prose-mirror plugin to generate decorations (in `Plugin.apply`). With the code as it currently exists, the transaction is applied to the plugin before the async operation completes. How do I trigger the transaction when `getAsyncData` completes? The `dispatch` operation does not seem to be doing the trick.

**Comment by tasercake** [comment]
In the same boat here — I have an async function that returns some data that I'd like to render using a node view.

I would also like a keyboard shortcut to do the same thing. So my code looks like:

```typescript
const MyNode = Node.create({
  name: 'myNode',
  addCommands() {
    return {
      asyncCommand: (getContent) => ({ chain }) => {
        getContent().then((data) => {
		  // This doesn't work
		  chain().insertContent(data).run()
		})
        return true
      }
    }
  },
  addKeyboardShortcuts() {
    return {
      Enter: ({ editor }) => {
        const contentGetter = async () => {
          // Await some async call here
          await new Promise(resolve => setTimeout(resolve, 1000))
          return '<p>some content</p>'
        }
        return editor.commands.asyncCommand(contentGetter)
      }
    }
  }
})
```
